internal/profiles/service.go

```diff
@@ -707,3 +743,58 @@ func deleteRuleStatusesForProfile(
 
 	return nil
 }
+
+func upsertRuleInstances(
+	ctx context.Context,
+	qtx db.Querier,
+	profileID uuid.UUID,
+	newRules []*minderv1.Profile_Rule,
+	entityType db.Entities,
+	rulesInProf RuleMapping,
+) ([]uuid.UUID, error) {
+	updatedIDs := make([]uuid.UUID, len(newRules))
+	for i, rule := range newRules {
+		// TODO: Clean up this logic once we no longer have to support the old tables.
+		entityRuleTuple, ok := rulesInProf[RuleTypeAndNamePair{
+			RuleType: rule.Type,
+			RuleName: rule.Name,
+		}]
+		if !ok {
+			return nil, fmt.Errorf("unable to find rule type ID for %s/%s", rule.Name, rule.Type)
+		}
+
+		def, err := json.Marshal(rule.Def)
+		if err != nil {
+			return nil, fmt.Errorf("unable to serialize rule def: %w", err)
+		}
+
+		params, err := json.Marshal(rule.Params)
+		if err != nil {
+			return nil, fmt.Errorf("unable to serialize rule params: %w", err)
+		}
+
+		newInstance := db.UpsertRuleInstanceParams{
+			ProfileID:  profileID,
+			RuleTypeID: entityRuleTuple.RuleID,
+			Name:       rule.Name,
+			EntityType: entityType,
+			Def: pqtype.NullRawMessage{
```

Comment:
The PR looks good, the only question I have is how would we later instantiate the rule params and defs - right now the code seems to presume that they are non-nil:
```
result, err = r.ingester.Ingest(ctx, inf.Entity, params.GetRule().Params.AsMap())
```
so we'll have to check that we don't instantiate them as `nil` in this case.